lib/nerves_hub_link/archive_manager.ex

```diff
@@ -0,0 +1,178 @@
+  end
+
+  @doc """
+  Returns the UUID of the currently downloading archive, or nil.
+  """
+  @spec currently_downloading_uuid(GenServer.server()) :: uuid :: String.t() | nil
+  def currently_downloading_uuid(manager \\ __MODULE__) do
+    GenServer.call(manager, :currently_downloading_uuid)
+  end
+
+  @doc false
+  @spec child_spec(map()) :: Supervisor.child_spec()
+  def child_spec(args) do
+    %{
+      start: {__MODULE__, :start_link, [args, [name: __MODULE__]]},
+      id: __MODULE__
+    }
+  end
+
+  @doc false
+  @spec start_link(GenServer.options()) :: GenServer.on_start()
+  def start_link(args, opts \\ []) do
+    GenServer.start_link(__MODULE__, args, opts)
+  end
+
+  @impl GenServer
+  def init(args) do
+    {:ok, %__MODULE__{data_path: args.data_path}}
+  end
+
+  @impl GenServer
+  def handle_call({:apply_archive, %ArchiveInfo{} = info}, _from, %__MODULE__{} = state) do
+    state = maybe_update_archive(info, state)
+    {:reply, state.status, state}
+  end
+
+  def handle_call(:currently_downloading_uuid, _from, %__MODULE__{archive_info: nil} = state) do
+    {:reply, nil, state}
+  end
+
+  def handle_call(:currently_downloading_uuid, _from, %__MODULE__{} = state) do
+    {:reply, state.archive_info.uuid, state}
+  end
+
+  def handle_call(:status, _from, %__MODULE__{} = state) do
+    {:reply, state.status, state}
+  end
+
+  @impl GenServer
+  def handle_info({:update_reschedule, response}, state) do
+    {:noreply, maybe_update_archive(response, %__MODULE__{state | update_reschedule_timer: nil})}
+  end
+
+  # messages from Downloader
+  def handle_info({:download, :complete}, state) do
+    Logger.info("[NervesHubLink] Archive Download complete")
+    _ = Client.archive_ready(state.archive_info, state.file_path)
+
+    {:noreply,
+     %__MODULE__{state | archive_info: nil, file_path: nil, download: nil, status: :idle}}
+  end
+
+  def handle_info({:download, {:error, reason}}, state) do
+    Logger.error("[NervesHubLink] Nonfatal HTTP download error: #{inspect(reason)}")
+    {:noreply, state}
+  end
+
+  # Data from the downloader
+  def handle_info({:download, {:data, data}}, state) do
+    :ok =
+      File.open!(state.file_path, [:append], fn fd ->
+        IO.binwrite(fd, data)
+      end)
+
+    {:noreply, state}
+  end
+
+  defp maybe_update_archive(info, state) do
+    # Cancel an existing timer if it exists.
+    # This prevents rescheduled updates`
+    # from compounding.
+    state = maybe_cancel_timer(state)
+
+    uri = URI.parse(info.url)
+    file_name = Path.basename(uri.path)
+    file_path = Path.join(state.data_path, "archives/#{file_name}")
+    directory = Path.dirname(file_path)
+
+    pid = self()
+
+    case Client.archive_available(info) do
+      :download ->
+        {:ok, download} = Downloader.start_download(info.url, &send(pid, {:download, &1}))
+
+        _ = File.mkdir_p(directory)
+        # delete the old one in case it was a failed download
+        _ = File.rm_rf(file_path)
```

Comment:
Should we be more clever about this? If things did fail to download, we now don't even have the old archive. Maybe we need to download to a `${filepath}.download` and then copy it over to `filepath` when successfully completed?